src/main/java/org/apache/commons/io/output/DeferredFileOutputStream.java

```diff
@@ -316,8 +330,10 @@ public void writeTo(final OutputStream out) throws IOException
         }
 
         if (isInMemory()) {
+            assert memoryOutputStream != null : "@AssumeAssertion(nullness): isInMemory returns true when threshold is not exceeded; memoryOutputStream is reset to null after threshold is reached.";
```

Comment:
I don't think this is true.
It's possible for a call to `checkThreshold()` to switch the stream from in memory to on disk, without changing the values that `isInMemory()` and `isThresholdExceeded()` depend upon.
Please make such a test case; I think you will be able to cause a NullPointerException.

Reply:
Yes, I missed checking condition based on `written` and `threshold` fields. I will work on this part and inform you.

Reply:
I tried many test-cases with different values of `written`, `threshold`, and `count`. In each test-case results were `Ok` without any exception.

 `written+count > threshold` is necessary condition to call `thresholdReached()` in `checkThreshold()`. After executing this method, the important operation is `written = written + count`. Doing this balances the return value of `isInMemory()` and value of `memoryOutputStream`.      
 
I think that `NPE` wont occur in this case.

Reply:
I think you are right that there will not be a `NPE`.  I stand by my claim that `checkThreshold()` can violate the invariant.  However, `checkThreshold()` is protected, and every method that calls `checkThreshold()` re-establishes the invariant.  I had not noticed that `checkThreshold()` was protected when I wrote my first comment.  Thanks for investigating this!